test/CodeGenILTests/IndexingMicroTests.cs

```diff
@@ -0,0 +1,217 @@
+b = 5;
+c=a[b];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            Assert.AreEqual(5, output.Values.ToList()[2][0]);
+        }
+        [Test]
+        public void IndexIdentArrayWithIdent_Doubles()
+        {
+            var dscode = @"
+import(""DesignScriptBuiltin.dll"");
+a = [0.0,1.1,2.0,3.0,4.0,5.0];
+b = 5;
+c=a[b];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            Assert.AreEqual(5.0, output.Values.ToList()[2][0]);
+        }
+        [Test]
+        public void IndexIdentArrayWithIdent_MixedElementTypeArray()
+        {
+            var dscode = @"
+import(""DesignScriptBuiltin.dll"");
+a = [0.0,1.1,2.0,3.0,4.0,5.0,6,7,8,9,10];
+b = 1;
+c=a[b];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            Assert.AreEqual(1.1, output.Values.ToList()[2][0]);
+        }
+        [Test]
+        [Category("Failure")]//fails because we need to emit multiple ldelem opcodes per index.
+        public void IndexArrayWithArrayConstants()
+        {
+            var dscode = @"
+        import(""DesignScriptBuiltin.dll"");
+        a = [0,1,2,3,4,5];
+        b=a[[0,5]];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            CollectionAssert.AreEqual(new long[] { 0, 5 }, output.Values.ToList()[1][0] as long[]);
+        }
+        [Test]
+        public void IndexNestedArray_WithIdent()
+        {
+            var dscode = @"
+import(""DesignScriptBuiltin.dll"");
+a = [[0,1,2],[3,4,5]];
+b = 1;
+c = 0;
+d=a[b][c];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            Assert.AreEqual(3, output.Values.ToList()[3][0]);
+        }
+
+        [Test]
+        [Category("Failure")]//this fails because we can't index into an object
+                             //and [0..10] is assumed to be an object[] because it's generated with a function call.
+                             //this test actaully then fallsback correctly to ValueAtIndex but calls the wrong overload.
+        public void IndexIntoNestedRange_WithIdent()
+        {
+            var dscode = @"
+import(""DesignScriptBuiltin.dll"");
+a = [0..10]; //a is a []object
+b = 0;
+c = 5;
+d=a[b][c];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            Assert.AreEqual(4, output.Values.ToList()[3][0]);
+        }
+
+        [Test]
+        [Category("Failure")]//to accomplish this either need to generate a function call, or emit conditional il.
```

Comment:
For constant arrays, can you not get the length, and subtract the -ve index from the length to get the actual index?

Reply:
@aparajit-pratap yes, that is possible if we know all values ahead of time, but if the value is an identifier then we don't currently have the capability to see if it's negative - instead what I propose is to emit MSIL which tests if the value is negative at runtime, this will make indexing cost a few more operations each but we can still avoid the cost of a function call.

Another approach is to use a try/catch and fallback to negative indexing in the catch block.
please see:
https://jira.autodesk.com/browse/DYN-5154

Reply:
Can this be solved using your idea of not just caching types in the `variables ` dictionary but also values of the identifiers in it?

Reply:
I think that would help solve the issue partially - but it won't help when the index is returned from a function call - let's say we have a program like:
```
a = [1,2,3,4];
b = getRandomInt(-1,1);
c = a[b];
```
by looking at the types, we will decide that we can compile this index operation, we can't know what the value is, and half the time we'll generate index calls that will throw at runtime.

We can either:
1. implement index IL that checks if the value at runtime is negative 
2. stop supporting negative indexing.
3. don't generate IL index ops unless we index using constants.
4. only perform indexing by using function calls (get rid of this PR)